### PR TITLE
Modernize old loc reduction kernel tests.

### PR DESCRIPTION
**Pull request by rchen20**
# Summary

- This PR is a refactoring.
- It converts the old kernel loc reduction tests to use the new testing framework. To do:

- [x]   - Explore using Camp resource with array-of-pointers.
- [x]   - Add other loc tests.
- [x]   - Add TBB, OMP.
Added Cuda and Hip as well.
- [ ]   - Combine min and max loc tests into one test filtered by type?
Cancelled: Tests would become too unclear.
- [x]   - Combine CUDA versions of tests with the standard loc kernel test?
- [ ]   - Extract types commonly used between reduce-basic and region tests?
Cancelled: Not enough commonality.

### CODE REVIEW

File: test/functional/kernel/reduce-basic/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+if(RAJA_ENABLE_OPENMP)
+  list(APPEND KERNEL_LOC_BACKENDS OpenMP)
+endif()
+
+if(RAJA_ENABLE_TBB)
+  list(APPEND KERNEL_LOC_BACKENDS TBB)
+endif()
+
+
+#
```
Comment (rhornung67) [inline test/functional/kernel/reduce-basic/CMakeLists.txt:21]: why are the tests broken apart into two different subsets, one for CPU back-ends and one for GPU back-ends? We should strive to drive each test code with all beck-ends.

Comment (rchen20) [reply test/functional/kernel/reduce-basic/CMakeLists.txt:21]: The non-tuple tests only work on the CPU backends because the GPU backends require a local tuple object for indexing. The ViewTuple tests can adequately perform 2D indexing on the device because they use a local RAJA::make_tuple.

I will try adding the omp-target loc-reduce tests to functional/forall/, but there are none existing which use RAJA::kernel.

Comment (rchen20) [reply test/functional/kernel/reduce-basic/CMakeLists.txt:21]: I suppose I can eliminate the custom 2D indexing used in the non-tuple CPU backend tests, or copy that object to the working resource.

Comment (rhornung67) [reply test/functional/kernel/reduce-basic/CMakeLists.txt:21]: Please try to make the CPU and GPU tests as close to the same as possible. If there are differences, that indicates that we may have some work to do. Thanks.

Comment (rchen20) [reply test/functional/kernel/reduce-basic/CMakeLists.txt]: Turns out those tests were already working for GPU and OpenMPTarget . . .

**Comment by rhornung67** [comment]
@rchen20 let me know when this is ready for final review. Thanks.

**Comment by rchen20** [comment]
> @rchen20 let me know when this is ready for final review. Thanks.

@rhornung67 Will do. I want to add a couple more tests, but they will follow the same structure as these new tests.

**Comment by rchen20** [comment]
Hello @trws @davidbeckingsale, the errors I'm getting in this PR are coming from CAMP's tuple_out_of_range test. It is operating correctly, and passing, but the older GCCs (and one version of clang) are also erroring during compilation (e.g. https://travis-ci.com/github/LLNL/RAJA/jobs/452191206). The compilation error is expected, but it's causing the CI system to report failure. Is there some way around this?

File: test/old-tests/unit/cuda/test-reduce-randloc.cpp [hunk 2]
```diff
@@ -1,463 +0,0 @@
-//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
-// Copyright (c) 2016-20, Lawrence Livermore National Security, LLC
-// and RAJA project contributors. See the RAJA/COPYRIGHT file for details.
-//
-// SPDX-License-Identifier: (BSD-3-Clause)
-//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
-
-///
-/// Source file containing tests for RAJA GPU tuple min-loc reductions
```
Comment (rchen20) [inline test/old-tests/unit/cuda/test-reduce-randloc.cpp:9]: @rhornung67 I did not duplicate these tests because most of them are needlessly complex, and their most important functionality is already covered in the forall/indexset tests. Let me know if this is ok with you.

**Comment by rchen20** [comment]
@MrBurmark @rhornung67 @davidbeckingsale This should be ready to go, would you mind taking a look at this when you get chance? Thanks!

File: test/include/RAJA_test-reduceloc-data.hpp [hunk 3]
```diff
@@ -0,0 +1,42 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
```
Comment (rhornung67) [inline test/include/RAJA_test-reduceloc-data.hpp:1]: @rchen20 you could use the methods in RAJA_test-forall-data.hpp in the test include directory and just not use the "test_array" to keep the code smaller.

Comment (rchen20) [reply test/include/RAJA_test-reduceloc-data.hpp:1]: I was thinking about that; I can if I add a remove_ref to the type of the first argument in allocateforalldata, e.g. something like remove_ref<typename T> N. The indirection arrays (workarr2D, checkarr2D) are causing this to be the wrong type.

File: test/functional/kernel/reduce-basic/test-kernel-loc.cpp.in [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+  >;
+
+#if defined(RAJA_ENABLE_OPENMP)
+
+using OpenMPKernelLocExecPols =
+  camp::list<
+
+    RAJA::KernelPolicy<
+      RAJA::statement::For<1, RAJA::omp_parallel_for_exec,  // row
+        RAJA::statement::For<0, RAJA::omp_parallel_for_exec,  // col
```
Comment (MrBurmark) [inline test/functional/kernel/reduce-basic/test-kernel-loc.cpp.in:70]: Do we want to have nested parallelism here?

Comment (rchen20) [reply test/functional/kernel/reduce-basic/test-kernel-loc.cpp.in:70]: It doesn't fail when run, but I can make the outer loop a RAJA::seq_exec.

Comment (rhornung67) [reply test/functional/kernel/reduce-basic/test-kernel-loc.cpp.in:70]: or you could use a loop collapse policy....maybe not the best for performance, but it would be a good test.

File: test/include/RAJA_test-reduceloc-data.hpp (hunk 3 above)
Comment (rchen20) [reply test/include/RAJA_test-reduceloc-data.hpp]: @rhornung67 Changed this to use forall-data.

File: test/functional/kernel/reduce-basic/test-kernel-loc.cpp.in (hunk 4 above)
Comment (rchen20) [reply test/functional/kernel/reduce-basic/test-kernel-loc.cpp.in]: @MrBurmark @rhornung67 Modified these policies for use of collapse and seq_exec.

Comment (MrBurmark) [reply test/functional/kernel/reduce-basic/test-kernel-loc.cpp.in]: I'm happy with that, should we add collapse to omp target?

Comment (rchen20) [reply test/functional/kernel/reduce-basic/test-kernel-loc.cpp.in]: There doesn't appear to be a corresponding collapse policy for omp_target.

Comment (rhornung67) [reply test/functional/kernel/reduce-basic/test-kernel-loc.cpp.in]: Yes, there is. It's used in many kernels in the RAJA Perf Suite; e.g., https://github.com/LLNL/RAJAPerf/blob/develop/src/basic/NESTED_INIT-OMPTarget.cpp#L67

Comment (rchen20) [reply test/functional/kernel/reduce-basic/test-kernel-loc.cpp.in]: Added it in now. Thanks!

File: test/include/RAJA_test-forall-data.hpp [hunk 5]
```diff
@@ -15,7 +15,7 @@
 #include "camp/resource.hpp"
 
 template<typename T>
-void allocateForallTestData(T N,
+void allocateForallTestData(camp::type::ptr::rem<T> N,
```
Comment (MrBurmark) [inline test/include/RAJA_test-forall-data.hpp:18]: If this is just a size, why wasn't this size_t to begin with?

File: test/functional/kernel/reduce-basic/test-kernel-loc.cpp.in (hunk 4 above)
Comment (rchen20) [reply test/functional/kernel/reduce-basic/test-kernel-loc.cpp.in]: @rhornung67 I've commented out the omp_target_parallel_collapse_exec test for the moment. It always produces the wrong answer, most likely because the data is not being copied over to the device. Does this work correctly in the PerfSuite?

Comment (rhornung67) [reply test/functional/kernel/reduce-basic/test-kernel-loc.cpp.in]: All of the PerfSuite kernels that use the omp target collapse stuff (RAJA::kernel) yield correct results. However, none of those kernels use reductions. There are several kernels that use reductions with RAJA::forall and those produce correct results for omp target. There may be a missing 'map' clause in the code someplace.

**Comment by rchen20** [comment]
@trws @rhornung67 I created a PR in CAMP https://github.com/LLNL/camp/pull/46 which should help solve some of these TOSS-gcc6 Travis errors. For some reason I cannot add reviewers to the PR, so if you could take a quick look at it that would be great. Thanks!

**Comment by rchen20** [comment]
LGTM

**Comment by rhornung67** [comment]
@rchen20 is this PR ready for a final review?

**Comment by rchen20** [comment]
@rhornung67 I was going to address the "size_t N" issue that @MrBurmark had mentioned, although that would involve other tests as well. Otherwise, the rest of this is PR ready.

**Comment by rhornung67** [comment]
OK. Please let me know when you are done making changes.

**Comment by rchen20** [comment]
@rhornung67 @MrBurmark This should be ready to go. The input sizes of test cases have been reduced to fit what OpenMP target can handle. AllocateForallData has been overloaded as well, but functionality remains the same.

**Review by MrBurmark — APPROVED** [review]

**Review by rhornung67 — APPROVED** [review]
looks good. 👍
